Move type checking methods out of basic.py

samdoran opened the pull request:
##### SUMMARY
Move the following type checking methods out of `basic.py` into `lib/ansible/module_utils/common/validation.py` as standalone functions:
- `check_type_str()`
- `check_type_list()`
- `safe_eval()`
- `check_type_dict()`
- `check_type_bool()`
- `check_type_int()`
- `check_type_float()`
- `check_type_path()`
- `check_type_raw()`
- `check_type_bytes()`
- `check_type_bits()`
- `check_type_jsonarg()`

Move the following out of `basic.py` and into `lib/ansible/module_utils/common/text/formatters.py`:
- `lenient_lowercase()`
- `human_to_bytes()`
- `bytes_to_human()`

Move the following out of `basic.py` and into `lib/ansible/module_utils/common/text/converters.py`:
- `jsonify()`
- `json_dict_to_unicode_to_bytes()` --> `container_to_bytes()`
- `json_dict_bytes_to_unicode()` --> `container_to_text()`

Create common code for testing `json` imports in `lib/ansible/module_utils/common/_json_compat.py`

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
`lib/ansible/module_utils/common/_json_compat.py`
`lib/ansible/module_utils/common/validation.py`
`lib/ansible/module_utils/basic.py`

samdoran committed:
Move check_type_str() out of basic.py

samdoran committed:
Move check_type_list() out of basic.py

samdoran committed:
Move check_type_dict() out of basic.py

sivel commented on lib/ansible/module_utils/basic.py:
```diff
@@ -179,6 +167,11 @@
+    from ansible.module_utils.common._json_compat import json
+except ImportError as e:
+    print('\n{{"msg": "Error: ansible requires the stdlib json {0}", "failed": true}}'.format(to_native(e)))
+    sys.exit(1)
```
This import should likely be moved back where the original code was, so that it trips early, and other places that import `_json_compat` don't cause pre-mature exceptions that aren't handled

samdoran replied:
Ok. It needs to be after `to_native()` which is why I put it down here, to avoid having to move that import. But I'll do the needful.

sivel commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -9,6 +9,10 @@
+try:
+    from ansible.module_utils.common._json_compat import json
+except Exception:
+    raise
```
The `try/except` here isn't needed, just import.

samdoran committed:
Move json importing code to common location

samdoran committed:
Move check_type_bool() out of basic.py

samdoran committed:
Move _check_type_int() out of basic.py

samdoran committed:
Move _check_type_float() out of basic.py

samdoran commented:
bot_skip

samdoran committed:
Move _check_type_path() out of basic.py

samdoran committed:
Move _check_type_raw() out of basic.py

samdoran committed:
Move _check_type_bytes() out of basic.py

samdoran committed:
Move _check_type_bits() out of basic.py

samdoran committed:
Create text.formatters.py

Move human_to_bytes, bytes_to_human, and _lenient_lowercase out of basic.py into text.formatters.py
Change references in modules to point to function at new location

samdoran committed:
Move _check_type_jsonarg() out of basic.py

Also move JSON realated functions to module_utils.json_utils

samdoran committed:
Rename json related functions and put them in common.text.converters

Move formatters.py to common.text.formatters.py and update references in modules.

samdoran committed:
Move safe_eval() out of basic.py

# Conflicts:
#	lib/ansible/module_utils/common/validation.py

# Conflicts:
#	lib/ansible/module_utils/common/validation.py

samdoran committed:
Order imports

samdoran committed:
Rework check_type_str()

Add allow_conversion option to make the function more self-contained.
Move the messaging back to basic.py since those error messages are more relevant to using this function in the context of AnsibleModule and not when using the function in isolation.

Akasurde commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -281,3 +294,213 @@ def check_missing_parameters(module_parameters, required_parameters=None):
+
+def check_type_bytes(value):
+    try:
+        human_to_bytes(value)
```
```suggestion
        return human_to_bytes(value)
```

Akasurde commented on lib/ansible/module_utils/common/validation.py:
```diff
@@ -281,3 +294,213 @@ def check_missing_parameters(module_parameters, required_parameters=None):
+
+def check_type_bits(value):
+    try:
+        human_to_bytes(value, isbits=True)
```
```suggestion
        return human_to_bytes(value, isbits=True)
```

samdoran committed:
Add unit tests for type checking functions

Akasurde commented on test/units/module_utils/common/validation/test_check_type_str.py:
```diff
@@ -0,0 +1,33 @@
+)
+
+
+def test_check_type_str():
```
```suggestion
@pytest.mark.parametrize('test_value, expected_value', TEST_CASES)
def test_check_type_str(test_value, expected_value):
    assert expected_value == check_type_str(test_value)
```

Akasurde commented on lib/ansible/module_utils/basic.py:
```diff
@@ -1740,160 +1603,59 @@ def _check_argument_values(self, spec=None, param=None):
+                msg = common_msg.capitalize()
+                raise TypeError(to_native(msg))
+            elif self._string_conversion_action == 'warn':
+                msg = ('The value {0!r} (type {0.__class__.__name__}) in a string field was converted to {1!r} (type string). '
```
Can we add a `key` name as well ? It is very confusing message if we have multiple warnings like this - 

```
[WARNING]: The value 600 (type int) in a string field was converted to u'600' (type string). If this does not look like what you expect,
quote the entire value to ensure it does not change.

 [WARNING]: The value {'Name': 'aws-blorp_rhel7'} (type dict) in a string field was converted to u"{'Name': 'aws-blorp_rhel8'}" (type
string). If this does not look like what you expect, quote the entire value to ensure it does not change.

 [WARNING]: The value 300 (type int) in a string field was converted to u'300' (type string). If this does not look like what you expect,
quote the entire value to ensure it does not change.
```

Maybe something like - 

```suggestion
                msg = ('The value {0!r} (type {0.__class__.__name__}) in a string field of argument {1} was converted to {2!r} (type string). '
```

samdoran replied:
I'm not sure of a good way to do this since we don't have the name of the `key` available, only the value to check or convert.

Akasurde replied:
Got it. But this is very confusing error message. anyways, not a blocker as such.

samdoran replied:
I figured the value would be enough to track down the problematic field. Even if you have multiple fields with the same value, the error only prints once (a byproduct of having the same error message text).

But I see that as a warning, it gets printed at the top of the play and that removes a lot of context. 🤔 As an `error`, it's much more clear what is happening.

There is room for improvement here, for sure.

sivel commented on lib/ansible/module_utils/common/text/formatters.py:
```diff
@@ -0,0 +1,100 @@
+}
+
+
+def _lenient_lowercase(lst):
```
Since we are going to import this back into basic.py, it probably shouldn't be "private"

samdoran committed:
Change _lenient_lowercase to lenient_lowercase per feedback
